Bytesize nftest transcript (#1512)
nf-core/nf-co.re: markdown/events/2022/bytesize_nftest.md

<details markdown="1"><summary>Video transcription</summary>
**Note: The content has been edited for reader-friendliness**

[0:01](https://www.youtube.com/watch?v=K9B7JRkMpQ4&t=1)
(host) Hello everyone, welcome to today's bytesize. I'm Franziska Bonath, and I'm the host for today. With us is Edmund Miller, and he is going to give us his impressions of working with nf-test.

[0:16](https://www.youtube.com/watch?v=K9B7JRkMpQ4&t=16)
Okay. Good morning, everybody. Before we start, I did not write nf-test, I've just been using it. This was actually written by Lucas Forer and Sebastian Schoenherr, I'm going to butcher that from iMed. Okay, so what is nf-test? It is basically the ability to write unit tests for Nextflow workflows, which I looked at Phil's talk, is the second most requested feature in this year's Nextflow and nf-core community survey. So you can think of it as Pytest, but for Nextflow, if you're familiar with Python or unit tests, basically just a way to write either full pipeline tests or sub-workflow function tests and module tests, and we'll get into some of those.

[1:14](https://www.youtube.com/watch?v=K9B7JRkMpQ4&t=74)

> So what that looks like. If you haven't memorized the Hello Nextflow script, here's a quick reminder. We start saying all of these different greetings in different languages, and we pipe them through, and then we view it lastly. And so what that looks like on the nf-test side is then we have a name of the test, this spoiler plate up here, and then we have a script, and we're actually calling, I believe, the remote script here, and then a name of the test on this line, and then we have what we expect, such as the success of it, how many tasks we're going to have, which I think is also important as you're changing things, and then we have several certs of what's happening in the standard out of these greetings, for example. And that's the power of what you can do with nf-test here, and I'll let your imagination run wild with that of all the things you can do with it from there.

Reviewer: ```suggestion
So what that looks like. If you haven't memorized the Hello Nextflow script, here's a quick reminder. We start saying all of these different greetings in different languages, and we pipe them through, and then we view it lastly. And so what that looks like on the nf-test side is then we have a name of the test, this boiler plate up here, and then we have a script, and we're actually calling, I believe, the remote script here, and then a name of the test on this line, and then we have what we expect, such as the success of it, how many tasks we're going to have, which I think is also important as you're changing things, and then we have several asserts of what's happening in the standard out of these greetings, for example. And that's the power of what you can do with nf-test here, and I'll let your imagination run wild with that of all the things you can do with it from there.
```

[9:50](https://www.youtube.com/watch?v=K9B7JRkMpQ4&t=590)
This is just some boilerplate. It boxes each workflow into its own separate execution directory, I believe. Phil and I have had some interesting findings with that. Again, you can check the size. Now what is the really powerful part here is this snapshot feature. What this does is, it basically pulls in the md5sums for you of all of these files. So we have things like metrics, the run manifest here. We have our fastqs that are zipped up and you can specify all these different paths. This actually isn't a perfect example, because you can just list a directory, I found that really powerful. You can just point it there and then it'll go through all the files. What that looks like over here is that you get this snapshot file.

[10:40](https://www.youtube.com/watch?v=K9B7JRkMpQ4&t=640)

> What this does is it just creates a quick little snapshot with the md5sums of all those files that you specified. And you can do standard out, standard error, et cetera as well on those based on the workflow. But we found that to be a little hairy for right now. As you can see, this is automated. And whenever you rerun your test, it checks for these md5sums and will tell you. The other thing is it's really easy when you know you changed some stuff or you know these are going to change. You just run it with the `-update_snapshots` flag and it updates all the snapshots for you automatically. Then you just commit this file and it will check again against them.

Reviewer: ```suggestion
What this does is it just creates a quick little snapshot with the md5sums of all those files that you specified. And you can do standard out, standard error, et cetera as well on those based on the workflow. But we found that to be a little hairy for right now. As you can see, this is automated. And whenever you rerun your test, it checks for these md5sums and will tell you. The other thing is it's really easy when you know you changed some stuff or you know these are going to change. You just run it with the `--update-snapshot` flag and it updates all the snapshots for you automatically. Then you just commit this file and it will check again against them.
```

[11:58](https://www.youtube.com/watch?v=K9B7JRkMpQ4&t=718)
Let's talk about some good testing practices as we start to adopt this and start to adopt more testing. I know we have great testing practices in the nf-core tools. So there's things like test driven development and that's where you write your test first and then you write your code after. There's also acceptance driven test driven development. That's where you start to write your acceptance criteria first and then you write your tests and do your development. There's behavior driven development, where you start to talk to the team first and then you go through those and start to write your tests and then do your development after that. There's also pair programming, which is where you write the code, someone else writes the tests, you come together and work on it together. Lastly... just kidding, we're not going to talk about any of those today.

[12:57](https://www.youtube.com/watch?v=K9B7JRkMpQ4&t=777)

> Those are all great and lots of great things that you should go read about. But I think as a community, we have more important things to focus on. What are some realistic testing priorities? We should probably first convert our CI tests to pipeline tests. I think those are great to reproduce locally and quickly. Next, we can then add more pipeline tests for various params and pathways that you might want to make sure don't break, I think most people are just testing that all their liners work and then hope that everything downstream works, whereas you might want to check all of your skips, for example, and make sure that you get the right amount of processes and that certain things aren't getting run.

Reviewer: ```suggestion
Those are all great and lots of great things that you should go read about. But I think as a community, we have more important things to focus on. What are some realistic testing priorities? We should probably first convert our CI tests to pipeline tests. I think those are great to reproduce locally and quickly. Next, we can then add more pipeline tests for various params and pathways that you might want to make sure don't break, I think most people are just testing that all their aligners work and then hope that everything downstream works, whereas you might want to check all of your skips, for example, and make sure that you get the right amount of processes and that certain things aren't getting run.
```

[13:44](https://www.youtube.com/watch?v=K9B7JRkMpQ4&t=824)
Another great practice that I hope that we can adopt is to add tests when fixing bugs in PRs that prove that this bug is fixed and we're not going to come back to this bug and checking for it in the future so that we don't end up recreating it. Lastly, I think testing your local modules can go a long way. I think we have a lot of custom scripts that we could do some tests on and make sure that we're getting the expected output out of those to verify our scientific results. So the TLDR is let's avoid regressions and move forward with our tests.

[14:22](https://www.youtube.com/watch?v=K9B7JRkMpQ4&t=862)

> A couple of us have been toying around a lot how a roll out plan could look like. This is just a quick update on this and is no way indicative of exactly what's going to happen. First we want to start with pipelines as a proving ground because I think that's where we had the most ground to gain compared to being stuck with just GitHub actions. So it's already in methyl-seq and demultiplexed. If you're looking for some quick examples and you want to implement this today and switch those over, just that step one, we're going to need to update the template. I've got a PR started for that, but we need to settle on some best practices or how we want to do things so that as you switch between nf-core pipelines and contributing, things are familiar.

Reviewer: ```suggestion
A couple of us have been toying around a lot how a roll out plan could look like. This is just a quick update on this and is no way indicative of exactly what's going to happen. First, we want to start with pipelines as a proving ground because I think that's where we had the most ground to gain compared to being stuck with just GitHub actions. So it's already in methyl-seq and demultiplex. If you're looking for some quick examples and you want to implement this today and switch those over, just that step one, we're going to need to update the template. I've got a PR started for that, but we need to settle on some best practices or how we want to do things so that as you switch between nf-core pipelines and contributing, things are familiar.
```
